sockeye/layers.py

```diff
@@ -46,11 +48,1542 @@ def activation(data: mx.sym.Symbol, act_type: str) -> mx.sym.Symbol:
+    """
+
+    @abstractmethod
+    def create_encoder_layer(self, input_num_hidden: int, prefix: str) -> EncoderLayer:
+        pass
+
+    @abstractmethod
+    def create_decoder_layer(self, input_num_hidden: int, prefix: str) -> DecoderLayer:
+        pass
+
+
+class SharedEncoderDecoderLayer(DecoderLayer, EncoderLayer):
+
+    @abstractmethod
+    def process_sequence(self,
+                         data: mx.sym.Symbol,
+                         lengths: mx.sym.Symbol,
+                         max_length: int) -> Tuple[mx.sym.Symbol, mx.sym.Symbol, int]:
+        """
+        Process either the encoder or the decoder hidden states.
+        :param data: Encoded source layer: (source_encoded_max_length, batch_size, encoder_depth).
+        :param lengths: Lengths of hidden sequences. Shape: (batch_size,).
+        :param max_length: Maximum length.
+        :return: Encoded versions of the input data, the new sequence lenghts and the new maximum length.
+        """
+        pass
+
+    def encode_sequence(self, source_encoded: mx.sym.Symbol, source_encoded_lengths: mx.sym.Symbol,
+                        source_encoded_max_length: int, att_dict) -> Tuple[mx.sym.Symbol, mx.sym.Symbol, int]:
+        return self.process_sequence(source_encoded, source_encoded_lengths, source_encoded_max_length)
+
+    def decode_sequence(self,
+                        source_encoded: mx.sym.Symbol,
+                        source_encoded_lengths: mx.sym.Symbol,
+                        source_encoded_max_length: int,
+                        target_encoded: mx.sym.Symbol,
+                        target_encoded_lengths: mx.sym.Symbol,
+                        target_encoded_max_length: int,
+                        target_autoregressive_bias: mx.sym.Symbol) -> mx.sym.Symbol:
+        new_target_encoded, new_target_encoded_lengths, new_target_encoded_max_length = self.process_sequence(
+            target_encoded, target_encoded_lengths, target_encoded_max_length)
+        assert new_target_encoded_lengths is target_encoded_lengths, C.SEQUENCE_LENGTH_MUST_NOT_CHANGE_MSG
+        assert new_target_encoded_max_length == target_encoded_max_length, C.SEQUENCE_LENGTH_MUST_NOT_CHANGE_MSG
+        return new_target_encoded
+
+
+def layers_with_states_iter(layers: List[DecoderLayer],
+                            step: int,
+                            layer_states_flat: List[mx.sym.Symbol]) -> Iterator[Tuple[DecoderLayer, List[mx.sym.Symbol]]]:
+    state_idx = 0
+    for layer in layers:
+        if layer.num_states(step) != 0:
+            layer_states = layer_states_flat[state_idx:state_idx + layer.num_states(step)]
+            state_idx += layer.num_states(step)
+        else:
+            layer_states = []
+
+        yield layer, layer_states
+
+
+class NestedDecoderLayer(DecoderLayer):
+    """
+    A decoder layer which combines several other decoder sub-layers.
+    """
+
+    @property
+    @abstractmethod
+    def layers(self) -> List[DecoderLayer]:
```

Comment:
maybe rename layers to sublayers in this class? Its not like a chain but more like a parent layer right?